### PR TITLE
update PUPPI based on DP note

### PR DESCRIPTION
**Pull request by nhanvtran**
This is the companion PR to #10551

### CODE REVIEW

**Comment by nhanvtran** [comment]
This supercedes #10404

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
ping

**Comment by nhanvtran** [comment]
re-indenting updated

File: CommonTools/PileupAlgos/plugins/PuppiProducer.cc [hunk 1]
```diff
@@ -100,43 +116,103 @@ void PuppiProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
...
-      
-      if(closestVtx != 0 && pVtxId == 0 && fabs(pReco.charge) > 0) pReco.id = 1;
-      if(closestVtx != 0 && pVtxId >  0 && fabs(pReco.charge) > 0) pReco.id = 2;
-      //Add a dZ cut if wanted (this helps)
-      if(fUseDZ && pDZ > -9999 && closestVtx == 0 && (fabs(pDZ) < fDZCut) && fabs(pReco.charge) > 0) pReco.id = 1;
-      if(fUseDZ && pDZ > -9999 && closestVtx == 0 && (fabs(pDZ) > fDZCut) && fabs(pReco.charge) > 0) pReco.id = 2;
-      
-      //std::cout << "pVtxId = " << pVtxId << ", and charge = " << itPF->charge() << ", and closestVtx = " << closestVtx << ", and id = " << pReco.id << std::endl;
-      //std::cout << "charge = " << itPF->charge() << ", pDZ = " << pDZ << ", pVtxId = " << pVtxId << ", closestVtx = " << closestVtx << ", fromPV() = " << lPack->fromPV() << ", pReco.id = " << pReco.id << std::endl;
+    //   //std::cout << "pVtxId = " << pVtxId << ", and charge = " << itPF->charge() << ", and closestVtx = " << closestVtx << ", and id = " << pReco.id << std::endl;
```
Comment (slava77) [inline CommonTools/PileupAlgos/plugins/PuppiProducer.cc]: are these debugs still really needed?

File: CommonTools/PileupAlgos/src/PuppiAlgo.cc [hunk 2]
```diff
@@ -151,6 +154,8 @@ void PuppiAlgo::computeMedRMS(const unsigned int &iAlgo,const double &iPVFrac) {
     double lAdjust = double(lNPV)/double(fPupsPV.size()+fNCount[iAlgo]);
     if(lAdjust > 0) fMedian[iAlgo] -= sqrt(ROOT::Math::chisquared_quantile(lAdjust,1.)*fRMS[iAlgo]);
 
+    // std::cout << "*change2* -- iAlgo = " << iAlgo << ", fRMSEtaSF = " << fRMSEtaSF << ", fMedEtaSF = " << fMedEtaSF << ", fRMS[iAlgo] = " << fRMS[iAlgo] << ", fMedian[iAlgo] = " << fMedian[iAlgo] << std::endl;
```
Comment (slava77) [inline CommonTools/PileupAlgos/src/PuppiAlgo.cc]: only commented out couts and empty lines have changed. Were these changes really necessary?

File: CommonTools/PileupAlgos/test/testPUMods.py [hunk 3]
```diff
@@ -5,37 +5,41 @@
 process.load('FWCore/MessageService/MessageLogger_cfi')
 process.load('Configuration/StandardSequences/FrontierConditions_GlobalTag_cff')
 process.MessageLogger.cerr.FwkReport.reportEvery = 10
-from Configuration.AlCa.GlobalTag import GlobalTag
-process.GlobalTag = GlobalTag(process.GlobalTag, 'auto:run2_mc', '')
+process.GlobalTag.globaltag = 'START53_V7G::All'
```
Comment (slava77) [inline CommonTools/PileupAlgos/test/testPUMods.py]: is this a working config?
GT points to 53, the file seems to be from 74X.

File: CommonTools/PileupAlgos/test/testPUMods.py [hunk 4]
```diff
@@ -5,37 +5,41 @@
...
 process.load('CommonTools/PileupAlgos/Puppi_cff')
 process.load('CommonTools/PileupAlgos/softKiller_cfi')
 
-process.maxEvents = cms.untracked.PSet( input = cms.untracked.int32(50) )
+process.maxEvents = cms.untracked.PSet( input = cms.untracked.int32(10) )
 process.source = cms.Source("PoolSource",
-                            fileNames  = cms.untracked.vstring('/store/relval/CMSSW_7_5_0_pre4/RelValQCD_FlatPt_15_3000HS_13/MINIAODSIM/MCRUN2_75_V1-v1/00000/328B19C2-4DF6-E411-92C1-003048FFD796.root')
+	fileNames  = cms.untracked.vstring(
+		# '/store/mc/RunIISpring15DR74/QCD_Pt-15to7000_TuneCUETP8M1_Flat_13TeV_pythia8/GEN-SIM-RECO/AsymptFlat0to50bx25Reco_MCRUN2_74_V9-v3/10000/0009D30B-0207-E511-B581-0026182FD753.root'
+		'file:/uscms_data/d2/ntran/physics/Jets/PUPPITuning2015/JMEValidatorWorkbench/cmssw748_pr_Aug3/src/25202.0_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25--old/step3_inMINIAODSIM.root'
```
Comment (slava77) [inline CommonTools/PileupAlgos/test/testPUMods.py]: private files should not be in the src code. Is there a cms-wide available file?

**Comment by slava77** [comment]
@cmsbuild please test
except for the commented out code, changes seem OK now.
@arizzi  @gpetruc  please check the PackedCandidateProducer is OK from your side

**Comment by slava77** [comment]
Looks like PuppiWeightNoLep_ is not initialized

File: CommonTools/PileupAlgos/test/testPUMods.py (hunk 4 above)
Comment (nhanvtran) [inline CommonTools/PileupAlgos/test/testPUMods.py]: done

File: CommonTools/PileupAlgos/test/testPUMods.py (hunk 3 above)
Comment (nhanvtran) [inline CommonTools/PileupAlgos/test/testPUMods.py]: reverted back to auto

File: CommonTools/PileupAlgos/src/PuppiAlgo.cc (hunk 2 above)
Comment (nhanvtran) [inline CommonTools/PileupAlgos/src/PuppiAlgo.cc]: removed

File: CommonTools/PileupAlgos/plugins/PuppiProducer.cc (hunk 1 above)
Comment (nhanvtran) [inline CommonTools/PileupAlgos/plugins/PuppiProducer.cc]: removed

**Comment by nhanvtran** [comment]
Sorry, in the great re-indenting I lost two lines.  They are back in now.  Also I fixed some line-by-liners above.

**Comment by gpetruc** [comment]
Re-indents look fine to me now.
Will test asap

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@nhanvtran please make a 76X PR as well

**Comment by slava77** [comment]
With this PR, CPU spent in puppi-related operations triples (from ~0.5s to 1.5s).
The total miniAOD CPU is now ~15% or reco in PU=35@25ns.

@gpetruc You will probably start getting even more comments like the recent ones on how to make packed cands without puppi

File: PhysicsTools/PatAlgos/plugins/PATPackedCandidateProducer.cc [hunk 5]
```diff
@@ -224,11 +233,33 @@ void pat::PATPackedCandidateProducer::produce(edm::Event& iEvent, const edm::Eve
...
+           float puppiWeightNoLepVal = 0.0;
+
+           // Check the "no lepton" puppi weights. 
+           // If present, then it is not a lepton, use stored weight
+           // If absent, it is a lepton, so set the weight to 1.0
+           if ( puppiWeightNoLep.isValid() ) {
+             // Look for the pointer inside the "no lepton" candidate collection.
+             auto foundNoLep = puppiCandsNoLep->end();
+             for ( auto s = puppiCandsNoLep->begin(),  s_end = puppiCandsNoLep->end(); s != s_end; ++s ) {       
+               if ( s->sourceCandidatePtr(0) == pkref->sourceCandidatePtr(0) ) {
```
Comment (slava77) [inline PhysicsTools/PatAlgos/plugins/PATPackedCandidateProducer.cc]: This line increases packedPFCandidates by 20 times.
Does this really need to go through the sourceCandidatePtr?
If needed, computation of sourceCandidatePtr(0) should be moved out of this nested loop and done outside of the loop over cands. A vector of Ptrs can then be compared in this nested loop. (x1000 speedup).
Also, this loop should be terminated after the first match.

Comment (slava77) [inline PhysicsTools/PatAlgos/plugins/PATPackedCandidateProducer.cc]: https://slava77sk.web.cern.ch/slava77sk/reco/cgi-bin/igprof-navigator/CMSSW_7_5_X_2015-08-18-2300-sign575.25202.0.step5.pp/104

**Comment by slava77** [comment]
I mentioned the other offender, PuppiContainer::var_within_R , in the past
https://slava77sk.web.cern.ch/slava77sk/reco/cgi-bin/igprof-navigator/CMSSW_7_5_X_2015-08-18-2300-sign575.25202.0.step5.pp/68
Different code should be used to collect a set of candidates in a dR cone.

this should be fixed by the time we go to 76X. Even now, with this PR the re-miniAOD will be ~2 times slower due to puppi modules.

**Comment by slava77** [comment]
http://media3.s-nbcnews.com/i/streams/2013/January/130124/1C5708530-tdy_roker_dog_120912.jpg

**Comment by nhanvtran** [comment]
thanks @slava77 ;)

I'm looking at the choke points you noted above.  Generally we have some other ideas to speed up the code as well but given the miniAOD time constraints were going to add it to the "final" PR.  we can see how much we can add now...

**Comment by nhanvtran** [comment]
sorry, by "miniAOD time constraints" i meant getting the PRs in -- not the speed of the processing..

**Comment by slava77** [comment]
On 8/19/15 8:37 AM, nhanvtran wrote:
> sorry, by "miniAOD time constraints" i meant getting the PRs in -- not
> the speed of the processing..
> 

yeah, time constraints everywhere.

I think the PackedCandidate issue is easy to change, hence suggested a
fix ~now.
For the deltaR, more code may need to be moved around. So, that could
come later
when we expect to run on more events than in the scope of the re-miniAOD

**Comment by nhanvtran** [comment]
@slava77 is this what you had in mind?

File: PhysicsTools/PatAlgos/plugins/PATPackedCandidateProducer.cc [hunk 6]
```diff
@@ -121,6 +125,11 @@ void pat::PATPackedCandidateProducer::produce(edm::Event& iEvent, const edm::Eve
     iEvent.getByToken( PuppiCands_, puppiCands );
     std::vector<int> mappingPuppi(puppiCands->size());
 
+    edm::Handle< edm::ValueMap<float> > puppiWeightNoLep;
+    iEvent.getByToken( PuppiWeightNoLep_, puppiWeightNoLep );
+    edm::Handle<std::vector< reco::PFCandidate > > puppiCandsNoLep;
+    iEvent.getByToken( PuppiCandsNoLep_, puppiCandsNoLep );    
+
```
Comment (slava77) [inline PhysicsTools/PatAlgos/plugins/PATPackedCandidateProducer.cc]: add here 
```c++
std::vector<reco::CandidatePtr> puppiCandsNoLepPtrs;
if (puppiCandsNoLep.isValid(){
for (auto pup : *puppiCandsNoLep){
puppiCandsNoLepPtrs.push_back(pup->sourceCandidatePtr(0));
}
```
then use it below in the loop

File: PhysicsTools/PatAlgos/plugins/PATPackedCandidateProducer.cc [hunk 7]
```diff
@@ -224,11 +233,35 @@ void pat::PATPackedCandidateProducer::produce(edm::Event& iEvent, const edm::Eve
...
+           // Check the "no lepton" puppi weights. 
+           // If present, then it is not a lepton, use stored weight
+           // If absent, it is a lepton, so set the weight to 1.0
+           if ( puppiWeightNoLep.isValid() ) {
+             // Look for the pointer inside the "no lepton" candidate collection.
+             auto foundNoLep = puppiCandsNoLep->end();
+             auto pkrefPtr = pkref->sourceCandidatePtr(0);
+             for ( auto s = puppiCandsNoLep->begin(),  s_end = puppiCandsNoLep->end(); s != s_end; ++s ) {       
+               if ( s->sourceCandidatePtr(0) == pkrefPtr ) {
```
Comment (slava77) [inline PhysicsTools/PatAlgos/plugins/PATPackedCandidateProducer.cc]: this is not enough

**Comment by nhanvtran** [comment]
ok @slava77, have both fixes in now

File: PhysicsTools/PatAlgos/plugins/PATPackedCandidateProducer.cc [hunk 8]
```diff
@@ -224,11 +240,35 @@ void pat::PATPackedCandidateProducer::produce(edm::Event& iEvent, const edm::Eve
...
+           // If present, then it is not a lepton, use stored weight
+           // If absent, it is a lepton, so set the weight to 1.0
+           if ( puppiWeightNoLep.isValid() ) {
+             // Look for the pointer inside the "no lepton" candidate collection.
+             auto pkrefPtr = pkref->sourceCandidatePtr(0);
+             bool foundNoLep = false;
+             for ( size_t ipcnl = 0; ipcnl < puppiCandsNoLepPtrs.size(); ipcnl++){
+              if (puppiCandsNoLepPtrs[ipcnl] == pkrefPtr){
+                foundNoLep = true;
+                  puppiWeightNoLepVal = puppiCandsNoLepPtrs[ipcnl]->pt()/cand.pt(); // a hack for now, should use the value map
```
Comment (slava77) [inline PhysicsTools/PatAlgos/plugins/PATPackedCandidateProducer.cc]: the previous code was picking pt from the puppiCandsNoLep, not from the source ptr.
If it matters functionally, you should probably have 
```c++
// define outside the loop: auto const& puppiWeightNoLepV = puppiWeightNoLep.product();
puppiWeightNoLepVal = puppiWeightNoLepV[ipcnl].pt()/cand.pt();
```

**Comment by nhanvtran** [comment]
Ok @slava77  updated

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/10597	06dcff6
- code changes are OK
- jenkins tests pass and comparisons with baseline show no differences
- additionally tested in CMSSW_7_5_X_2015-08-18-2300 /test area sign575/ with PAT step only running on  70 events input from ttbar PU=35@25ns
    - CPU increase is now more acceptable: 
        - PuppiProducer went up from by 35% (two producers are running now), which is achieved by removal of PuppiContainer::getRawAlphas calls (previously 40% of the PuppiProducer)
        - pat::PATPackedCandidateProducer increase in CPU is now only 70% (from x20 in the last iteration) https://slava77sk.web.cern.ch/slava77sk/reco/cgi-bin/igprof-navigator/CMSSW_7_5_X_2015-08-18-2300-sign575.06dcff6.25202.0.step5.pp/595 vs https://slava77sk.web.cern.ch/slava77sk/reco/cgi-bin/igprof-navigator/CMSSW_7_5_X_2015-08-18-2300-orig.25202.0.step5.pp/807
            - here a large fraction of the increase seems to be from reading the additional collections and some manipulations. 
            - Also, the compiler decided to inline the PackedCandidate constructor, which makes the comparison difficult.
    - compared to baseline the number of puppi jets is smaller (as expected, see #10404 ). packed candidates get smaller, but this may be incidental 
```
-------------------------------------------------------------
 or, B       new, B      delta, B   delta, %   deltaJ, %    branch 
-------------------------------------------------------------
    19054 ->       18904       -150     -0.8  -0.21     patPackedCandidates_packedPFCandidates__PAT.
     2723 ->        2291       -432    -17.2  -0.60     patJets_slimmedJetsPuppi__PAT.
```
